### ISSUE TITLE
Add Glossary to dashboard

### ISSUE THREAD

**Issue by JIMENOFONSECA** [architecture-building-systems/CityEnergyAnalyst#1956]
Until now, @Jack-Hawthorne created a great glossary of all the variables inside CEA and organized them in here https://city-energy-analyst.readthedocs.io/en/latest/glossary.html

The process is semi-automated

Now, it would be good that the user can have access to a "SEARCH" function inside the dashboard, which could help him/her to find the meaning and units of any variable.

@reyery any ideas on how to integrate this into the dashboard?

**Comment by Jack-Hawthorne** [comment]
@reyery @JIMENOFONSECA 
the glossary is currently out of date

Back end
----------
i have some ideas and am currently working out a script that will merge the plots naming.csv to include attributes such as description, unit etc...

 to create a formal 'variable declaration' which can be verified using the meta_data created by trace_inputlocator

specifically, by referencing the trace data and the variable declaration, we will be able to see which variables have actually been written given a scenario, which are documented in variable declaration and which aren't, and then delete out-dated or include the new variables each time they are changed.

in this way we will be able to automatically update the glossary.

Front end
-----------
the search feature can then just reference the variable declaration (updated naming.csv in plots)

**Comment by Jack-Hawthorne** [comment]
@daren-thomas do you happen to have a complete trace.inputlocator.output.yml for the singapore cooling case?

**Comment by daren-thomas** [comment]
@Jack-Hawthorne no, I'm afraid not. This is blocked by #1958. I think it's possible to proceed without having all the data, no? The official meta-data is located in `cea/schemas.yml` - pending #1992, which introduces it along with the `cea.scripts.schemas()` function.

**Comment by reyery** [comment]
@JIMENOFONSECA currently the data I am pulling for the front end only contains variables for plots, using the `naming.csv` file. Would you like to include other variables such as those for input/output databases? @Jack-Hawthorne is there anywhere I could extract this other information?

**Comment by Jack-Hawthorne** [comment]
@reyery so i have in mind a re-vamp of the naming.csv to include all written variables, not just those for plots. my idea so far is to:

- cross reference with the meta-data from trace-inputlocator, checking to see if all variables in the naming.csv are ACTUALLY used (i have a feeling there are many outdated ones in there)
- for the ones which appear in the naming.csv, merge the naming.csv variables with the existing glossary (this old glossary can then be deleted)
- check to see if any are used but NOT in the naming.csv (i.e. unknown variables)
- add these to the naming.csv to have a comprehensive and complete list.
- add attributes to the variables in naming.csv: SCRIPT,VARIABLE,DESCRIPTION,VALUES,TYPE,COLOR from the meta-data
- make this merging process somewhat automated and easy between meta-data and naming.csv

Status:
i have currently already written a script to do the merging, however @daren-thomas i have not fully tested everything and would need the FULL up-to-date metadata before creating a PR. this is to avoid any losses in information which could result in extra work, such as the stripping the old glossary.

this will be a bit of a hassle, however, once completed should be a nice easy (enough) way to check that all the written variables have been declared. 

@JIMENOFONSECA @daren-thomas if you both have any input or can foresee some redundancy with your work on the schema for SQL, it would be much appreciated.

**Comment by daren-thomas** [comment]
@JIMENOFONSECA, are the variables in `naming.csv` always output fields from scripts?

@Jack-Hawthorne can you please create a sketch / mockup of the output you expect to create? we have an idea of the `schemas.yml` file (though, the list of columns could be a list instead of a dictionary - i'll look into that should it arise) and we already have the `naming.csv`. Which ones are you proposing to change?

**Comment by JIMENOFONSECA** [comment]
The variables in naming.csv where created to account for variables that we include in the plots of CEA.

There might be more variables that are not yet part of the plots of CEA, which are parts of the outputs.

To test changes in Naming, I will 1. run all the plots of CEA. I mean all the 60, and check that we did not erased any variable from naming.csv.

I think naming.csv has the variables we need to produce all plots, but it does not have all the input variables and more output variables of the rest of the scripts.




On 23 May 2019, 3:31 PM +0800, Daren Thomas <notifications@github.com>, wrote:

@JIMENOFONSECA<https://github.com/JIMENOFONSECA>, are the variables in naming.csv always output fields from scripts?

@Jack-Hawthorne<https://github.com/Jack-Hawthorne> can you please create a sketch / mockup of the output you expect to create? we have an idea of the schemas.yml file (though, the list of columns could be a list instead of a dictionary - i'll look into that should it arise) and we already have the naming.csv. Which ones are you proposing to change?

**Comment by Jack-Hawthorne** [comment]
> @JIMENOFONSECA, are the variables in `naming.csv` always output fields from scripts?
> 
> @Jack-Hawthorne can you please create a sketch / mockup of the output you expect to create? we have an idea of the `schemas.yml` file (though, the list of columns could be a list instead of a dictionary - i'll look into that should it arise) and we already have the `naming.csv`. Which ones are you proposing to change? also, if you have such a script already, can we see the output? even if the data is incomplete?

@daren-thomas 
ok, currently the documentation merge method works like this:
- read the old glossary file, read the naming.csv, read the metadata from trace script
- compare the metadata variables to those within naming.csv and glossary.csv
- if matched in only naming, parse the description from naming.csv (first priority) and put TODO for the other elements
- if matched in naming and glossary, parse description from naming and the other elements from glossary
- if matched in only glossary, parse all elements from glossary and put black for COLOR
- if no match occurs, put TODO for everything (note: need to record these ones in a list or something).
-output merged naming.csv
-raise exceptions: "Oh, looks like these variables haven't been documented: x, y, z"
-raise potentially outdated variables: "Oh, looks like these variables were found in naming.csv but weren't found in the metadata: x, y, z"

the overall schema for naming.csv aims to look like this:

SCRIPT,VARIABLE,DESCRIPTION,VALUES,TYPE,UNITS,COLOR
data-helper,BIOGAS,"Biogas gas   restricted in the area. 0 = no, 1, yes","{0, 1}","bool","None","baby vomit green"
demand,COAL_hs0_kW,Nominal Coal   consumption due to space heating,{0.0...n},float,"kW","tickle me pink"
etc...

subsequent modifications after the successful merge:
- delete the glossary part of the script and the old glossary file (no longer needed)
- reference the 'schema.yml' instead of the trace output meta-data

the method is really messy atm (i'll work at cleaning it up) and looks like this:
```
def meta_naming_merge(meta_location, output):
    import pandas
    import yaml

    # create dataframe from variables_gloss.csv without duplicates
    gloss_path = os.path.join(os.path.dirname(cea.config.__file__), '../docs/variables_gloss.csv')
    gloss = pandas.read_csv(gloss_path, delimiter=';')
    gloss = gloss.set_index(gloss['VARIABLE'])
    gloss = gloss[~gloss.index.duplicated(keep='first')]

    # create dataframe from naming.csv
    NAMING_FILE_PATH = os.path.join(os.path.dirname(cea.config.__file__), 'plots/naming.csv')
    naming = pandas.read_csv(NAMING_FILE_PATH)
    naming = naming.set_index(naming['VARIABLE'])

    # create dataframe from trace_inputlocator.output.yml
    TRACE_PATH = os.path.join(os.path.dirname(meta_location), 'baseline\\outputs\\trace_inputlocator.output.yml')
    with open(TRACE_PATH, 'r') as db:
        TRACE = yaml.load(db)

    # set of unique variables in trace_inputlocator.output.yml
    TRACE_DATA = set()
    # dict containing variables mapped to scripts which wrote them
    SCRIPT = {}
    for index in TRACE:
        if not TRACE[index]['created_by']:
            scr = 'primary_input'
        else:
            scr = TRACE[index]['created_by'][0]
        for VAR in TRACE[index]['schema']:
            if VAR.find('srf') != -1:
                VAR = VAR.replace(VAR, 'srf0')
            if VAR.find('PIPE') != -1:
                VAR = VAR.replace(VAR, 'PIPE0')
            if VAR.find('NODE') != -1:
                VAR = VAR.replace(VAR, 'NODE0')
            if VAR.find('B0') != -1:
                VAR = VAR.replace(VAR, 'B01')
            if TRACE[index]['file_type'] == 'epw':
                VAR = None
            if TRACE[index]['file_type'] == 'xlsx':
                for var in TRACE[index]['schema'][VAR]:
                    TRACE_DATA.add(var)
                    SCRIPT[var] = scr
            else:
                TRACE_DATA.add(VAR)
                SCRIPT[VAR] = scr

    # cross reference all variables(in order of priority) with naming.csv, variables_gloss.csv for descriptions
    # cross reference all variables with variables_gloss.csv for TYPE and VALUES
    # if variable not found in either - TO DO labels
    csv = pandas.DataFrame(columns=['VARIABLE', 'SCRIPT', 'DESCRIPTION', 'TYPE', 'VALUES'])
    vars = []
    desc = []
    dtype = []
    values = []
    script = []
    for var in TRACE_DATA:
        script.append(SCRIPT[var])
        if var in list(naming['VARIABLE']):
            vars.append(var)
            desc.append(naming.loc[var]['SHORT_DESCRIPTION'])
            if var in list(gloss['VARIABLE']):
                dtype.append(gloss.loc[var]['DTYPE'])
                values.append(gloss.loc[var]['VALUES'])
            if var not in list(gloss['VARIABLE']):
                dtype.append('TODO')
                values.append('TODO')

        elif var in list(gloss['VARIABLE']):
            vars.append(var)
            desc.append(gloss.loc[var]['SHORT_DESCRIPTION'])
            dtype.append(gloss.loc[var]['DTYPE'])
            values.append(gloss.loc[var]['VALUES'])
        else:
            vars.append(var)
            desc.append('TODO')
            dtype.append('TODO')
            values.append('TODO')

    # assign to dataframe and write
    csv['VARIABLE'] = vars
    csv['DESCRIPTION'] = desc
    csv['TYPE'] = dtype
    csv['VALUES'] = values
    csv['SCRIPT'] = script
    csv = csv.sort_values(by=['SCRIPT', 'VARIABLE', 'VALUES'])
    csv.to_csv(output, columns=['SCRIPT', 'VARIABLE', 'DESCRIPTION', 'VALUES', 'TYPE'], index=False)
```
hope this answers your questions

**Comment by Jack-Hawthorne** [comment]
@JIMENOFONSECA @daren-thomas 
here is the current output of this method:
[example_csv_file.txt](https://github.com/architecture-building-systems/CityEnergyAnalyst/files/3216126/example_csv_file.txt)

**Comment by daren-thomas** [comment]
@Jack-Hawthorne wait a minute... I just looked at `example_csv_file.txt` and I don't really think the column "script" is correct. Shouldn't it be the filename instead? or the locator method? I think you're throwing away information here for no reason at all. but maybe I'm missing something?

**Comment by Jack-Hawthorne** [comment]
@daren-thomas yeah this is now old, please check out the branch #2017 
the script is handy for themes which are already present in the naming.csv under the attribute 'category' 

the `naming_new.csv` one has the following attributes:

```
SCRIPT, LOCATOR_METHOD, FILE_NAME, VARIABLE, DESCRIPTION, UNIT, VALUES, TYPE, COLOR
```

with the primary key easily made from combining the `file_name` with the `variable`

**Comment by daren-thomas** [comment]
@reyery and @daren-thomas will skype about this.
